datastore/flow_postgres.go

```diff
@@ -71,172 +67,231 @@ func NewFlowPostgres(lookup environment.Environmenter, updater flow.Updater) (*F
 		return nil, fmt.Errorf("creating connection pool: %w", err)
 	}
 
-	flow := FlowPostgres{pool: pool, updater: updater}
+	flow := FlowPostgres{pool: pool}
 
 	return &flow, nil
 }
 
+// Close closes the connection pool.
+func (p *FlowPostgres) Close() {
+	p.pool.Close()
+}
+
 // Get fetches the keys from postgres.
 func (p *FlowPostgres) Get(ctx context.Context, keys ...dskey.Key) (map[dskey.Key][]byte, error) {
-	uniqueFieldsStr, fieldIndex, uniqueFQID := prepareQuery(keys)
-
-	// For very big SQL Queries, split them in part
-	if len(fieldIndex) > maxFieldsOnQuery {
-		keysList := splitFieldKeys(keys)
-		result := make(map[dskey.Key][]byte, len(keys))
-		for _, keys := range keysList {
-			resultPart, err := p.Get(ctx, keys...)
-			if err != nil {
-				return nil, fmt.Errorf("get key list: %w", err)
-			}
+	conn, err := p.pool.Acquire(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("acquiring connection: %w", err)
+	}
+	defer conn.Release()
 
-			for k, v := range resultPart {
-				result[k] = v
-			}
+	return getWithConn(ctx, conn.Conn(), keys...)
+}
+
+func getWithConn(ctx context.Context, conn *pgx.Conn, keys ...dskey.Key) (map[dskey.Key][]byte, error) {
+	collectionIDs := make(map[string][]int)
+	for _, key := range keys {
+		if slices.Contains(collectionIDs[key.Collection()], key.ID()) {
+			continue
 		}
-		return result, nil
+
+		collectionIDs[key.Collection()] = append(collectionIDs[key.Collection()], key.ID())
 	}
 
-	sql := fmt.Sprintf(`SELECT fqid, %s from models where fqid = ANY ($1) AND deleted=false;`, uniqueFieldsStr)
+	keyValues := make(map[dskey.Key][]byte, len(keys))
+	for collection, ids := range collectionIDs {
+		// TODO: maybe only fetch id and requested keys
+		sql := fmt.Sprintf(`SELECT * FROM "%s" WHERE id = ANY ($1) `, collection)
 
-	rows, err := p.pool.Query(ctx, sql, uniqueFQID)
-	if err != nil {
-		return nil, fmt.Errorf("sending query: %w", err)
-	}
-	defer rows.Close()
+		rows, err := conn.Query(ctx, sql, ids)
+		if err != nil {
+			return nil, fmt.Errorf("sending query for %s: %w", collection, err)
+		}
+		defer rows.Close()
```

Comment:
That is wrong. In go, defer is only called, when the function exists, not when a for-loop exists.

It is probably ok, to move this line inside the forEachRow callback